arrays.qmd

```diff
@@ -5,105 +5,234 @@
+```{r}
+library(odin2)
+library(dust2)
+```
 
-* Need a good motivating example for an array - previously we used an age structured model
-* show example code
+## Example: an age structured SIR model {#sec-stochastic-age}
 
-```
-update(S) <- S - n_SI
-update(I) <- I + n_SI - n_IR
-update(R) <- R + n_IR
-```
+Let's take the simple stochastic SIR model from @sec-stochastic-sir and add age structure to it, with heterogeneous mixing between age groups
 
-```
-update(S[]) <- S[i] - n_SI[i]
-update(I[]) <- I[i] + n_SI[i] - n_IR[i]
-update(R[]) <- R[i] + n_IR[i]
+```{r}
+sir_age <- odin({
+  # Equations for transitions between compartments by age group
+  update(S[]) <- S[i] - n_SI[i]
+  update(I[]) <- I[i] + n_SI[i] - n_IR[i]
+  update(R[]) <- R[i] + n_IR[i]
+  
+  # Individual probabilities of transition:
+  p_SI[] <- 1 - exp(-lambda[i] * dt) # S to I
+  p_IR <- 1 - exp(-gamma * dt) # I to R
+  
+  # Calculate force of infection
+  ## age-structured contact matrix: m[i, j] is mean number of contacts an
+  ## individual in group i has with an individual in group j per time unit
+  m <- parameter()
+  ## here s_ij[i, j] gives the mean number of contacts and individual in group
+  ## i will have with the currently infectious individuals of group j
+  s_ij[, ] <- m[i, j] * I[j]
+  ## lambda[i] is the total force of infection on an individual in group i 
+  lambda[] <- beta * sum(s_ij[i, ])
+  
+  # Draws from binomial distributions for numbers changing between
+  # compartments:
+  n_SI[] <- Binomial(S[i], p_SI[i])
+  n_IR[] <- Binomial(I[i], p_IR)
+  
+  initial(S[]) <- S0[i]
+  initial(I[]) <- I0[i]
+  initial(R[]) <- 0
+  
+  # User defined parameters - default in parentheses:
+  S0 <- parameter()
+  I0 <- parameter()
+  beta <- parameter(0.0165)
+  gamma <- parameter(0.1)
+  
+  # Dimensions of arrays
+  N_age <- parameter()
+  dim(S0) <- N_age
+  dim(I0) <- N_age
+  dim(S) <- N_age
+  dim(I) <- N_age
+  dim(R) <- N_age
+  dim(n_SI) <- N_age
+  dim(n_IR) <- N_age
+  dim(p_SI) <- N_age
+  dim(m) <- c(N_age, N_age)
+  dim(s_ij) <- c(N_age, N_age)
+  dim(lambda) <- N_age
+})
 ```
 
-The odin code
+In the odin code above
 
-```
+```r
 update(S[]) <- S[i] - n_SI[i]
 ```
 
 becomes (approximately)
 
-```
+```r
 for (int i = 0; i < S_length; ++i) {
   update_S[i] = S[i] + n_SI[i];
 }
 ```
 
-* Don't use index variables (`i`, `j`, `k`, etc) on the left hand side
-* Use multiple lines to account for boundary conditions
-* Can crash the program if out of bounds (we have plans for this)
+so there is an implicit indexing by `i` on the LHS in that equation of the odin code, and the generated code will then loop over all values of `i`. 
 
-## An age structured SIR model
 
+## Another example: an age-structured SIR model with vaccination
+Let's take the above model and additionally add some vaccination to it.
```

Comment:
```suggestion

Let's take the above model and additionally add some vaccination to it.
```

For consistency (and because the above does not properly render in all markdown dialects, so it's good practice generally)